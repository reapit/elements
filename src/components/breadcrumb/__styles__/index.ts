import { styled } from '@linaria/react'

export const ElBreadCrumbItem = styled.a`
  font-weight: var(--breadcrumb-item-font-weight);
  border-radius: var(--breadcrumb-item-border-radius);
  font-size: var(--breadcrumb-item-font-size);
  color: var(--breadcrumb-item-font-color);
  margin-right: var(--breadcrumb-item-spacing);
`

export const ElBreadCrumbContainer = styled.div`
  display: flex;
  width: 100%;
`
