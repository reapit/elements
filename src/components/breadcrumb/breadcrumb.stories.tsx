import { BreadCrumb } from './index'

export default {
  title: 'BreadCrumb',
  component: BreadCrumb,
}

export const BasicUsage = {
  render: ({}) => (
    <BreadCrumb
      items={[
        {
          text: 'Home',
          onClick: () => console.log('Home clicked'),
        },
        {
          text: 'Level 1',
          onClick: () => console.log('1 clicked'),
        },
        {
          text: 'Level 2',
          onClick: () => console.log('2 clicked'),
        },
        {
          text: 'Level 3',
          onClick: () => console.log('3 clicked'),
        },
      ]}
    />
  ),
}

export const DefaultIndex = {
  render: ({}) => (
    <BreadCrumb
      defaultActiveIndex={3}
      items={[
        {
          text: 'Home',
          onClick: () => console.log('Home clicked'),
        },
        {
          text: 'Level 1',
          onClick: () => console.log('1 clicked'),
        },
        {
          text: 'Level 2',
          onClick: () => console.log('2 clicked'),
        },
        {
          text: 'Level 3',
          onClick: () => console.log('3 clicked'),
        },
      ]}
    />
  ),
}
