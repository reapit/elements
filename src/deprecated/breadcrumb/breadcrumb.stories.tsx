import { DeprecatedBreadCrumb } from './index'

export default {
  title: 'Deprecated/DeprecatedBreadCrumb',
  component: DeprecatedBreadCrumb,
}

export const BasicUsage = {
  render: ({}) => (
    <DeprecatedBreadCrumb
      defaultActiveIndex={3}
      items={[
        {
          text: 'Home',
          onClick: () => console.log('Home clicked'),
        },
        {
          text: 'Level 1',
          onClick: () => console.log('1 clicked'),
        },
        {
          text: 'Level 2',
          onClick: () => console.log('2 clicked'),
        },
        {
          text: 'Level 3',
          onClick: () => console.log('3 clicked'),
        },
      ]}
    />
  ),
}

export const DefaultIndex = {
  render: ({}) => (
    <DeprecatedBreadCrumb
      defaultActiveIndex={3}
      items={[
        {
          text: 'Home',
          onClick: () => console.log('Home clicked'),
        },
        {
          text: 'Level 1',
          onClick: () => console.log('1 clicked'),
        },
        {
          text: 'Level 2',
          onClick: () => console.log('2 clicked'),
        },
        {
          text: 'Level 3',
          onClick: () => console.log('3 clicked'),
        },
      ]}
    />
  ),
}
