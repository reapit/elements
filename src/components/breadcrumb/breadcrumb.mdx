import { Story, Canvas, Controls, Meta } from '@storybook/blocks'
import { BreadCrumb } from './index'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as BreadcrumbStories from './breadcrumb.stories'

<Meta of={BreadcrumbStories} />

# BreadCrumb

## Basic Usage

<Controls />

Breadcrumbs are used to indicate to the user their flow in the application and provide a navigation back step to previous pages.

<Canvas>
  <Story of={BreadcrumbStories.BasicUsage} />
</Canvas>

<RenderHtmlMarkup of={BreadcrumbStories.BasicUsage} />

## Default Index

By setting the `defaultActiveIndex` prop, you can open the Bread crumb at the correct level when you navigate.

<Canvas>
  <Story of={BreadcrumbStories.DefaultIndex} />
</Canvas>

<RenderHtmlMarkup of={BreadcrumbStories.DefaultIndex} />
