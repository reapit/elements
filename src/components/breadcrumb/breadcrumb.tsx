import React, { Dispatch, FC, HTMLAttributes, MouseEvent, SetStateAction, useState } from 'react'
import { ElBreadCrumbItem, ElBreadCrumbContainer } from './__styles__'
import { Icon } from '../icon'
import { elMr2 } from '../../styles/spacing'
import { FlexContainer } from '../layout'
import { handleKeyboardEvent } from '../../storybook/handle-keyboard-event'

export interface BreadCrumbItem extends HTMLAttributes<HTMLAnchorElement> {
  text: string
  onClick: () => void
}

export interface BreadCrumbProps extends HTMLAttributes<HTMLElement> {
  items: BreadCrumbItem[]
  defaultActiveIndex?: number
}

export const handleNext =
  (setActive: Dispatch<SetStateAction<number>>, onClick: () => void, index: number) =>
  (e?: MouseEvent<HTMLAnchorElement>) => {
    e?.preventDefault()
    setActive(index)
    onClick()
  }

export const BreadCrumb: FC<BreadCrumbProps> = ({ items, defaultActiveIndex = 0, ...rest }) => {
  const [active, setActive] = useState<number>(defaultActiveIndex)

  return (
    <ElBreadCrumbContainer {...rest}>
      {items.map(({ onClick, text }, index) => {
        if (index > active) return null

        return (
          <FlexContainer isFlexAlignCenter key={index}>
            {Boolean(index) && <Icon className={elMr2} icon="chevronRight" intent="default" fontSize="12px" />}
            <ElBreadCrumbItem
              aria-current={active === index ? 'page' : 'false'}
              role="button"
              aria-label={`Breadcrumb item hit return to navigate to ${text}`}
              tabIndex={0}
              onClick={handleNext(setActive, onClick, index)}
              onKeyDown={handleKeyboardEvent('Enter', handleNext(setActive, onClick, index))}
            >
              {text}
            </ElBreadCrumbItem>
          </FlexContainer>
        )
      })}
    </ElBreadCrumbContainer>
  )
}
