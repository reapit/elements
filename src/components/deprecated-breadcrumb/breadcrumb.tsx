import React, { Dispatch, FC, HTMLAttributes, MouseEvent, SetStateAction, useState } from 'react'
import { ElDeprecatedBreadCrumbItem, ElDeprecatedBreadCrumbContainer } from './__styles__'
import { DeprecatedIcon } from '../deprecated-icon'
import { elMr2 } from '../../styles/spacing'
import { FlexContainer } from '../layout'
import { handleKeyboardEvent } from '../../storybook/handle-keyboard-event'

export interface DeprecatedBreadCrumbItem extends HTMLAttributes<HTMLAnchorElement> {
  text: string
  onClick: () => void
}

/** @deprecated will be replaced by new v5 BreadcrumbProps */
export interface DeprecatedBreadCrumbProps extends HTMLAttributes<HTMLElement> {
  items: DeprecatedBreadCrumbItem[]
  defaultActiveIndex?: number
}

export const handleNext =
  (setActive: Dispatch<SetStateAction<number>>, onClick: () => void, index: number) =>
  (e?: MouseEvent<HTMLAnchorElement>) => {
    e?.preventDefault()
    setActive(index)
    onClick()
  }

/** @deprecated will be replaced by new v5 Breadcrumb */
export const DeprecatedBreadCrumb: FC<DeprecatedBreadCrumbProps> = ({ items, defaultActiveIndex = 0, ...rest }) => {
  const [active, setActive] = useState<number>(defaultActiveIndex)

  return (
    <ElDeprecatedBreadCrumbContainer {...rest}>
      {items.map(({ onClick, text }, index) => {
        if (index > active) return null

        return (
          <FlexContainer isFlexAlignCenter key={index}>
            {Boolean(index) && (
              <DeprecatedIcon className={elMr2} icon="chevronRight" intent="default" fontSize="12px" />
            )}
            <ElDeprecatedBreadCrumbItem
              aria-current={active === index ? 'page' : 'false'}
              role="button"
              aria-label={`Breadcrumb item hit return to navigate to ${text}`}
              tabIndex={0}
              onClick={handleNext(setActive, onClick, index)}
              onKeyDown={handleKeyboardEvent('Enter', handleNext(setActive, onClick, index))}
            >
              {text}
            </ElDeprecatedBreadCrumbItem>
          </FlexContainer>
        )
      })}
    </ElDeprecatedBreadCrumbContainer>
  )
}
