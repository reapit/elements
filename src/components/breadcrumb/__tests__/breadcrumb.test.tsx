import { MouseEvent } from 'react'
import { render } from '@testing-library/react'
import { BreadCrumb, handleNext } from '../breadcrumb'

describe('BreadCrumb', () => {
  it('should match a snapshot', () => {
    const wrapper = render(
      <BreadCrumb
        defaultActiveIndex={3}
        items={[
          {
            text: 'Home',
            onClick: () => console.log('Home clicked'),
          },
          {
            text: 'Level 1',
            onClick: () => console.log('1 clicked'),
          },
          {
            text: 'Level 2',
            onClick: () => console.log('2 clicked'),
          },
          {
            text: 'Level 3',
            onClick: () => console.log('3 clicked'),
          },
        ]}
      >
        Hover here
      </BreadCrumb>,
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('handleNext', () => {
  it('should set active index and call onClick', () => {
    const setActive = vi.fn()
    const onClick = vi.fn()
    const index = 2

    const curried = handleNext(setActive, onClick, index)

    const mockEvent = {
      preventDefault: vi.fn(),
    } as unknown as MouseEvent<HTMLAnchorElement>

    curried(mockEvent)

    expect(setActive).toHaveBeenCalledWith(index)
    expect(onClick).toHaveBeenCalled()
    expect(mockEvent.preventDefault).toHaveBeenCalled()
  })

  it('should handle when event is not provided', () => {
    const setActive = vi.fn()
    const onClick = vi.fn()
    const index = 2

    const curried = handleNext(setActive, onClick, index)

    curried()

    expect(setActive).toHaveBeenCalledWith(index)
    expect(onClick).toHaveBeenCalled()
  })
})
