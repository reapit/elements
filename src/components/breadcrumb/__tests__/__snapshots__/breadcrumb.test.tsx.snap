// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BreadCrumb > should match a snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mocked-styled-1 el-bread-crumb-container"
      >
        <div
          class="el-flex-container el-flex-align-center"
        >
          <a
            aria-current="false"
            aria-label="Breadcrumb item hit return to navigate to Home"
            class="mocked-styled-0 el-bread-crumb-item"
            role="button"
            tabindex="0"
          >
            Home
          </a>
        </div>
        <div
          class="el-flex-container el-flex-align-center"
        >
          <span
            class="el-mr2 el-intent-default el-icon"
            style="font-size: 12px;"
          >
            <svg
              fill="none"
              height="24"
              role="img"
              style="font-size: 12px;"
              title="Icon image with name chevronRight"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <a
            aria-current="false"
            aria-label="Breadcrumb item hit return to navigate to Level 1"
            class="mocked-styled-0 el-bread-crumb-item"
            role="button"
            tabindex="0"
          >
            Level 1
          </a>
        </div>
        <div
          class="el-flex-container el-flex-align-center"
        >
          <span
            class="el-mr2 el-intent-default el-icon"
            style="font-size: 12px;"
          >
            <svg
              fill="none"
              height="24"
              role="img"
              style="font-size: 12px;"
              title="Icon image with name chevronRight"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <a
            aria-current="false"
            aria-label="Breadcrumb item hit return to navigate to Level 2"
            class="mocked-styled-0 el-bread-crumb-item"
            role="button"
            tabindex="0"
          >
            Level 2
          </a>
        </div>
        <div
          class="el-flex-container el-flex-align-center"
        >
          <span
            class="el-mr2 el-intent-default el-icon"
            style="font-size: 12px;"
          >
            <svg
              fill="none"
              height="24"
              role="img"
              style="font-size: 12px;"
              title="Icon image with name chevronRight"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <a
            aria-current="page"
            aria-label="Breadcrumb item hit return to navigate to Level 3"
            class="mocked-styled-0 el-bread-crumb-item"
            role="button"
            tabindex="0"
          >
            Level 3
          </a>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mocked-styled-1 el-bread-crumb-container"
    >
      <div
        class="el-flex-container el-flex-align-center"
      >
        <a
          aria-current="false"
          aria-label="Breadcrumb item hit return to navigate to Home"
          class="mocked-styled-0 el-bread-crumb-item"
          role="button"
          tabindex="0"
        >
          Home
        </a>
      </div>
      <div
        class="el-flex-container el-flex-align-center"
      >
        <span
          class="el-mr2 el-intent-default el-icon"
          style="font-size: 12px;"
        >
          <svg
            fill="none"
            height="24"
            role="img"
            style="font-size: 12px;"
            title="Icon image with name chevronRight"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <a
          aria-current="false"
          aria-label="Breadcrumb item hit return to navigate to Level 1"
          class="mocked-styled-0 el-bread-crumb-item"
          role="button"
          tabindex="0"
        >
          Level 1
        </a>
      </div>
      <div
        class="el-flex-container el-flex-align-center"
      >
        <span
          class="el-mr2 el-intent-default el-icon"
          style="font-size: 12px;"
        >
          <svg
            fill="none"
            height="24"
            role="img"
            style="font-size: 12px;"
            title="Icon image with name chevronRight"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <a
          aria-current="false"
          aria-label="Breadcrumb item hit return to navigate to Level 2"
          class="mocked-styled-0 el-bread-crumb-item"
          role="button"
          tabindex="0"
        >
          Level 2
        </a>
      </div>
      <div
        class="el-flex-container el-flex-align-center"
      >
        <span
          class="el-mr2 el-intent-default el-icon"
          style="font-size: 12px;"
        >
          <svg
            fill="none"
            height="24"
            role="img"
            style="font-size: 12px;"
            title="Icon image with name chevronRight"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.5045 20.0041C8.23904 20.0056 7.98389 19.9013 7.79542 19.7141C7.60635 19.5263 7.5 19.2707 7.5 19.0041C7.5 18.7375 7.60635 18.4819 7.79542 18.2941L14.0873 12.0041L7.79542 5.71409C7.4038 5.32197 7.4038 4.68621 7.79542 4.29409C8.18704 3.90197 8.82197 3.90197 9.21359 4.29409L16.2046 11.2941C16.3937 11.4819 16.5 11.7375 16.5 12.0041C16.5 12.2707 16.3937 12.5263 16.2046 12.7141L9.21359 19.7141C9.02511 19.9013 8.76996 20.0056 8.5045 20.0041Z"
              fill="currentColor"
            />
          </svg>
        </span>
        <a
          aria-current="page"
          aria-label="Breadcrumb item hit return to navigate to Level 3"
          class="mocked-styled-0 el-bread-crumb-item"
          role="button"
          tabindex="0"
        >
          Level 3
        </a>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
